pkg/streamingpromql/optimize/plan/narrow_selectors.go

```diff
@@ -64,16 +65,18 @@ func (n *NarrowSelectorsOptimizationPass) Apply(ctx context.Context, plan *plann
 	return plan, nil
 }
 
-func (n *NarrowSelectorsOptimizationPass) applyToNode(ctx context.Context, node planning.Node, res optimize.InspectResult) (bool, error) {
+// applyToNode attempts to recursively add hints to a binary expression node based on
+// fields being joined on or aggregated by. Returns true if any node had a hint added
+// to it.
+func (n *NarrowSelectorsOptimizationPass) applyToNode(ctx context.Context, node planning.Node) (bool, error) {
 	addedHint := false
 
-	switch e := node.(type) {
-	case *core.BinaryExpression:
-		// Set hints for a binary expression based on the expression itself and any
-		// children from the left hand side of the expression. Note that this stops
-		// after finding the first node that allows us to generate hints for the query
-		// (binary expressions or aggregations).
-		e.Hints = n.hintsFromNode(ctx, e, res)
+	if e, ok := node.(*core.BinaryExpression); ok {
+		// If this is a binary expression, try to find appropriate labels to use as hints
+		// based on joins or aggregations being performed by child nodes. We start with an
+		// empty "created" set of labels that we _cannot_ use as hints. This is populated
+		// and checked when generating hints.
+		e.Hints = n.hintsFromNode(ctx, e, nil)
```

Comment:
Realise this was here before, but we probably shouldn't overwrite `e.Hints`, as we might lose other hints (if there ever were any) - perhaps it would be safer to only set `Include`?